File: app/src/main/java/com/dreamer/mymy_moneytrack/ui/presenter/ShortSummaryPresenter.java
package com.dreamer.mymy_moneytrack.ui.presenter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import com.dreamer.mymy_moneytrack.MtApp;
import com.dreamer.mymy_moneytrack.R;
import com.dreamer.mymy_moneytrack.controller.FormatController;
import com.dreamer.mymy_moneytrack.entity.Period;
import com.dreamer.mymy_moneytrack.report.record.IRecordReport;
import com.dreamer.mymy_moneytrack.ui.presenter.base.BaseSummaryPresenter;

import java.text.SimpleDateFormat;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Util class to create and manage summary header view for .

 */
public class ShortSummaryPresenter extends BaseSummaryPresenter {

    @Inject
    FormatController formatController;

    private final int red;
    private final int green;
    private View view;

    public ShortSummaryPresenter(Context context) {
        this.context = context;
        MtApp.get().getAppComponent().inject(ShortSummaryPresenter.this);

        layoutInflater = LayoutInflater.from(context);
        red = context.getResources().getColor(R.color.red);
        green = context.getResources().getColor(R.color.green);
    }

    public interface ItemClickListener {
        void invoke();
    }

    public View create(boolean shortSummary, ItemClickListener itemClickListener) {
        view = layoutInflater.inflate(R.layout.view_summary_records, null);
        view.findViewById(R.id.iv_more).setVisibility(shortSummary ? View.VISIBLE : View.INVISIBLE);
        view.setEnabled(false);
        view.findViewById(R.id.lvSummary).setClickable(false);
        view.findViewById(R.id.cvSummary).setClickable(true);
        view.setTag(new ViewHolder(view, itemClickListener));

        return view;
    }

    public void update(IRecordReport report, String currency, List<String> ratesNeeded) {
        ViewHolder viewHolder = (ViewHolder) view.getTag();
        if (report == null) {
            viewHolder.tvTotalIncome.setText("");
            viewHolder.tvTotalExpense.setText("");

            viewHolder.tvTotal.setTextColor(red);
            viewHolder.tvTotal.setText(createRatesNeededList(currency, ratesNeeded));
        } else {
            viewHolder.tvPeriod.setText(formatPeriod(report.getPeriod()));

            viewHolder.tvTotalIncome.setTextColor(report.getTotalIncome() >= 0 ? green : red);
            viewHolder.tvTotalIncome.setText(formatController.formatIncome(report.getTotalIncome(),
                    report.getCurrency()));

            viewHolder.tvTotalExpense.setTextColor(report.getTotalExpense() > 0 ? green : red);
            viewHolder.tvTotalExpense.setText(formatController.formatExpense(report.getTotalExpense(),
                    report.getCurrency()));

            viewHolder.tvTotal.setTextColor(report.getTotal() >= 0 ? green : red);
            viewHolder.tvTotal.setText(formatController.formatIncome(report.getTotal(),
                    report.getCurrency()));
        }
    }

    private String formatPeriod(Period period) {
        switch (period.getType()) {
            case Period.TYPE_DAY:
                return period.getFirstDay();

            case Period.TYPE_MONTH:
                return new SimpleDateFormat("MMMM, yyyy").format(period.getFirst());

            case Period.TYPE_YEAR:
                return new SimpleDateFormat("yyyy").format(period.getFirst());

            case Period.TYPE_ALL_TIME:
                return context.getString(R.string.all_time);

            default:
                return context.getString(R.string.period_from_to, period.getFirstDay(),
                        period.getLastDay());
        }
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.tvPeriod)
        TextView tvPeriod;
        @BindView(R.id.tvTotalIncome)
        TextView tvTotalIncome;
        @BindView(R.id.tvTotalExpense)
        TextView tvTotalExpense;
        @BindView(R.id.tvTotal)
        TextView tvTotal;

        public ViewHolder(View view, final ItemClickListener itemClickListener) {
            super(view);
            ButterKnife.bind(this, view);
            view.findViewById(R.id.cvSummary).setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (itemClickListener != null)
                        itemClickListener.invoke();
                }
            });
        }
    }

}

File: app/src/main/java/com/dreamer/mymy_moneytrack/ui/presenter/ShortSummaryPresenter.kt
package com.dreamer.mymy_moneytrack.ui.presenter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import butterknife.ButterKnife
import com.dreamer.mymy_moneytrack.MtApp.Companion.get
import com.dreamer.mymy_moneytrack.R
import com.dreamer.mymy_moneytrack.controller.FormatController
import com.dreamer.mymy_moneytrack.entity.Period
import com.dreamer.mymy_moneytrack.report.record.IRecordReport
import com.dreamer.mymy_moneytrack.ui.presenter.base.BaseSummaryPresenter
import java.text.SimpleDateFormat
import javax.inject.Inject

/**
 * Util class to create and manage summary header view for .
 *
 */
class ShortSummaryPresenter(context: Context) : BaseSummaryPresenter() {

    var formatController: FormatController? = null
        @Inject set
    private val red: Int
    private val green: Int
    private var view: View? = null

    interface ItemClickListener {
        operator fun invoke()
    }

    fun create(shortSummary: Boolean, itemClickListener: (() -> Unit)?): View? {
        view = layoutInflater?.inflate(R.layout.view_summary_records, null)
        view?.findViewById<View>(R.id.iv_more)?.visibility
            ?: if (shortSummary) View.VISIBLE else View.INVISIBLE
        view?.isEnabled = false
        view?.findViewById<View>(R.id.lvSummary)?.isClickable = false
        view?.findViewById<View>(R.id.cvSummary)?.isClickable = true
        view?.tag = ViewHolder(view, itemClickListener)
        return view
    }

    fun update(report: IRecordReport?, currency: String?, ratesNeeded: List<String?>?) {
        val viewHolder = view!!.tag as ViewHolder
        if (report == null) {
            viewHolder.tvTotalIncome!!.text = ""
            viewHolder.tvTotalExpense!!.text = ""
            viewHolder.tvTotal!!.setTextColor(red)
            viewHolder.tvTotal!!.text = ratesNeeded?.let { createRatesNeededList(currency, it) }
        } else {
            viewHolder.tvPeriod!!.text = formatPeriod(report.period)
            viewHolder.tvTotalIncome!!.setTextColor(if (report.totalIncome >= 0) green else red)
            viewHolder.tvTotalIncome!!.text = formatController!!.formatIncome(
                report.totalIncome,
                report.currency
            )
            viewHolder.tvTotalExpense!!.setTextColor(if (report.totalExpense > 0) green else red)
            viewHolder.tvTotalExpense!!.text = formatController!!.formatExpense(
                report.totalExpense,
                report.currency
            )
            viewHolder.tvTotal!!.setTextColor(if (report.total >= 0) green else red)
            viewHolder.tvTotal!!.text = formatController!!.formatIncome(
                report.total,
                report.currency
            )
        }
    }

    private fun formatPeriod(period: Period): String {
        return when (period.type) {
            Period.TYPE_DAY -> period.firstDay
            Period.TYPE_MONTH -> SimpleDateFormat("MMMM, yyyy").format(period.first)
            Period.TYPE_YEAR -> SimpleDateFormat("yyyy").format(period.first)
            Period.TYPE_ALL_TIME -> context?.getString(R.string.all_time).toString()
            else -> context!!.getString(
                R.string.period_from_to, period.firstDay,
                period.lastDay
            )
        }
    }

    class ViewHolder(view: View?, itemClickListener: (() -> Unit)?) : RecyclerView.ViewHolder(
        view!!
    ) {
        @BindView(R.id.tvPeriod)
        var tvPeriod: TextView? = null

        @BindView(R.id.tvTotalIncome)
        var tvTotalIncome: TextView? = null

        @BindView(R.id.tvTotalExpense)
        var tvTotalExpense: TextView? = null

        @BindView(R.id.tvTotal)
        var tvTotal: TextView? = null

        init {
            ButterKnife.bind(this, view!!)
            view.findViewById<View>(R.id.cvSummary)
                .setOnClickListener { itemClickListener?.invoke() }
        }
    }

    init {
        this.context = context
        get()!!.appComponent!!.inject(this@ShortSummaryPresenter)
        layoutInflater = LayoutInflater.from(context)
        red = context.resources.getColor(R.color.red)
        green = context.resources.getColor(R.color.green)
    }
}